sql: make MVCC index creation option a metamorphic constant

stevendanna opened the pull request:
This makes the option `sql.mvcc_compliant_index_creation.enabled` a
metamorphic constant so that we ensure both paths are being tested.

Since the index creation scheme impacts IndexIDs and since not all
testing hooks are called along both paths, not all of the tests work
in both mode. Such tests have been modified to explicitly set the
setting to their expected value.

Release note: None

stevendanna committed:
sql: make MVCC index creation option a metamorphic constant

This makes the option `sql.mvcc_compliant_index_creation.enabled` a
metamorphic constant so that we ensure both paths are being tested.

Since the index creation scheme impacts IndexIDs and since not all
testing hooks are called along both paths, not all of the tests work
in both mode. Such tests have been modified to explicitly set the
setting to their expected value.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/78416)
<!-- Reviewable:end -->

stevendanna committed:
sql: allow zone configuration cleanup in tenants

Previously RemoveIndexZoneConfigs early-exited if it was run from a
tenant since zone configs weren't supported in tenants. Now, we do
support zone configurations in tenants and the early exit would result
in zone configurations not being removed when an index was dropped.

Release justification: Low risk bug fix for new functionality

Release note: None
